distribution/cachedir/manager.go

```diff
@@ -0,0 +1,214 @@
+	"github.com/docker/docker/api/types"
+	digest "github.com/opencontainers/go-digest"
+)
+
+// Manager handles the creation and deletion of cache directories and has a
+// very particular way of managing the cache. It was designed to be used by a
+// caller that manages the lifetime of the directories, but delegates the
+// mechanics of how to store and locate the directories to the Manager
+type Manager interface {
+	// GetDir creates a new cache dir or returns an existing one if one already
+	// exists for this key from a previous download. It also keeps track of how
+	// many references were made to the cache dir during
+	// ReleaseDir or DeleteDir have to be called exactly once per GetDir call
+	// and GetDir shouldn't be called again until one or the other is called.
+	GetDir(key string) string
+	// ReleaseDir lets the Manager know that the dir is no longer in use but
+	// should be kept around in case we need it again.
+	ReleaseDir(key string)
+	// DeleteDir lets the manager know that the dir is no longer used and can be
+	// deleted. If the delete operation fails and returns an error, the key is
+	// still considered released.
+	DeleteDir(key string) error
+	// CollectGarbage deletes all dirs that are not currently in use (ones that
+	// don't have a GetDir without a matching ReleaseDir/DeleteDir call) and
+	// returns the amount of space saved in bytes.
+	CollectGarbage() (uint64, error)
+	// Usage returns information about how much space the download cache is
+	// using. If any directories or files are unreadable, we just log the
+	// errors and don't count the files.
+	Usage() types.DownloadCacheUsage
+}
+
+type manager struct {
+	cacheRoot string
+	mu        sync.Mutex
+	inUse     map[digest.Digest]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[digest.Digest]struct{}{},
+	}
+}
+
+func (m *manager) cachePathForKeyHash(keyHash digest.Digest) string {
+	return filepath.Join(m.cacheRoot, keyHash.Algorithm().String(), keyHash.Hex())
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (m *manager) sizeEstimateByKeyHash(keyHash digest.Digest) uint64 {
+	dir := m.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.WithError(err).WithField("cachedir", dir).Warn("failed to determine size of directory")
+		return 0
+	}
+	return size
+}
+
+func (m *manager) cleanUpByKeyHash(keyHash digest.Digest) error {
+	return os.RemoveAll(m.cachePathForKeyHash(keyHash))
+}
+
+func (m *manager) GetDir(key string) string {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	hashedKey := digest.FromString(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debug("cachedir manager get")
```

Comment:
Maybe we should remove the `key` field, since it can be very large. Even for debug mode, the output ends up being very verbose. I see lots of messages like this:

```
DEBU[0434] cachedir manager get                          hashedKey=sha256:b848357614fa103de96fdf9127df65ae7e3634ad9b13ae98a8d20f00312bc3bf key="v2:sha256:5040bd2983909aa8896b9932438c3f1479d25ae837a5f6220242a264d0221f2dv2:sha256:568dce68541a55fd0fb33021383e3921145275a1f87d925f37f1246b682baf0bv2:sha256:6a832068e64c0a907e1585adac98802eb69cf9fc878ba00941d59e7be57c4b93v2:sha256:3b0f3d176a5bf080e41340e04575cdeece539b71cd0405812474a03151cdf73fv2:sha256:20cc248a5690169dac19b598857e4ecd4a6421a5c21eb699066d5006cf27834bv2:sha256:6ff565538ee684ae5f9d8e1c0c152bfabefa1f94202054df27edc82410555c74v2:sha256:9f1077228581cd19b5dc570b5831269295dd388feba6e783bd93396b0a825caav2:sha256:57359f144a1928936e5079e40f75cf1e00f8eb81b3bf2045fbdbc02e9f1b78e5v2:sha256:9754ef36b0333e14bbbc3bf78732d26f10b19d67c9f24cdc5f55885361df677bv2:sha256:e156df35b624fbec7438fceaa163977ebc6e9c72fed3d86cdfbe5068f8c50b56v2:sha256:df09daa2224a1b4c69e1c138218bbd96d9b929d5b4821ee81af8187ff2ac36f2v2:sha256:bfa0d803130201c1ead08ab4a858b17cf30be0281bb32ef87607379c24a063d5v2:sha256:58ec6fadb2c7f473eb1fdac86f56650a33d16203e5ebaddc4d101e6b9c1a8582v2:sha256:254919ae58d39c5f0f8700e27a39fd4c2b3f0f327a3da3ce844c3a95f1824a40v2:sha256:3fbda78d9add0fef30178b586bb24ab78eca7870380bae63f0b9724ff1c6f5f6v2:sha256:1f4006165e8995169156ffa99bf70faef3fdcfcbaa5daa22d69f39bd85887556v2:sha256:1989a57f5f0b764b3b21b95f547b6f1fa66c5d7d47c5c8cfc7d4091fcb41eda7"
```

Same goes for the other methods in this file that log in a similar way.

Alternatively we might just remove these log messages, since I'm not sure they are useful. Maybe it's good to keep the one for deletions, but I don't know that it makes sense to log whenever `GetDir` or `ReleaseDir` is called.